src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.LazyItemOperation.cs

```diff
@@ -232,48 +225,8 @@ protected void DecorateItemsWithMetadata(ICollection<I> items, ImmutableList<Pro
             /// </summary>
             protected ICollection<I> SelectItemsMatchingItemSpec(ImmutableList<ItemData>.Builder listBuilder, IElementLocation elementLocation)
```

Comment:
All the operations against one item type are collected into an immutable list. Then, when evaluation is done (or other scenarios) somebody calls [GetItems](https://github.com/Microsoft/msbuild/blob/master/src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs#L149-L180) on the list, which produces an immutable list of actual items. Each operation's apply thus gets a listbuilder to add / mutate items in. So far the common patterns for all the operations has been: produce some items (creation or selection), potentially do something to them, potentially save them back to the list builder.